docs/content/development/extensions-core/druid-basic-security.md

```diff
@@ -155,6 +189,27 @@ Example request body:
 }
 ```
 
+`GET(/druid-ext/basic-security/authentication/db/{authenticatorName}/config)`
+Return the authenticator current set of updatable configuration.
+
+`POST(/druid-ext/basic-security/authentication/db/{authenticatorName}/config)`
```

Comment:
Can you describe the intended use case for this dynamic config update and the dynamic credentials update for the Escalator?

Reply:
The use case for dynamic Escalator credentials update is to allow/support rolling update of  Escalator credentials within a Druid cluster without having to restart any services and risk any sort of services downtime. In a large scale production cluster if the Escalator credentials get comprised administrators will just need to dynamically update the credentials and not have to worry about restarting all the services within the cluster.
The same use case applies for dynamic authenticator config update. It will allow administrators to dynamically update ldap related configuration used to authenticating users without having to restart any services and create any downtime. The ldap configuration includes a bind password property, if the password gets compromised then it would be really nice if it can be updated across all services in rolling fashion without restarting any services and risk downtime

Reply:
Hm, I feel like this feature could be a bit tricky to use right, and I'm not sure its worth the added complexity in most cases.

If kept as a feature, I think users would need some guidance on how to do a successful credential migration in the docs.

Some points I think the docs should address:
- To migrate the credentials without downtime, the user would need to set up a temporary transition state where both the old and new credentials are valid (otherwise the credential update itself wouldn't be accepted by the cluster)
- Before pushing the new credentials to the cluster, the user should ensure that all of the configs on the cluster are updated to use the new credentials (so that they don't fall back to the old set upon restarts)
- After pushing the credential update, the user would need to check that the update was successfully propagated to all the cluster nodes (through the REST APIs? This would need a way to get the current escalator credentials or verify that they've changed in addition to the LDAP connection configs)
- To keep cluster config management simple, I think the in-metadata credentials should be treated as a temporary thing to assist with migration and not as a long-lived source of truth (which should be the config files). Adding DELETE endpoints for the in-metadata credentials and having the docs guide the user to delete that state after completing the migration would be good.
- I would move documentation for these config update APIs into a separate "credentials migration" section with the info above.

Reply:
Due to the complexity here I'd prefer to remove the feature as part of this codebase becoming an official Druid extension. To address downtime risk, we could / should recommend people upgrade clusters using a rolling update: http://druid.io/docs/latest/operations/rolling-updates.html. If a rolling update is done to replace compromised credentials, then as soon as it's finished, the old credentials can be revoked.

Reply:
Agree with gian, I think its ok to do the credentials upgrade via rolling-upgrades, 
same logic also applies to other secure configs like DB passwords, S3 access keys etc.

Reply:
I understand the concern for introducing additional complexity, and I’m open to removing this feature if that’s what everyone here seems to agree on. But to provide some perspective we have pretty large, and growing, cluster. We have about 70 historical nodes and with the number of segments we have on each historical node, our experience with restarting our historical cluster alone one node at a time is that it takes about 2 business days. Each of our historical node reads from SSD at max throughput of 2GB/s for about 2TB, so it takes about 16 minutes per node, at least, to restart that we spread over 2 business days. I agree this feature adds additional complexity but unless there’s a faster way of doing rolling updates, wouldn’t it make sense to be able to address any security related compromises within a cluster of significant size quickly that doesn’t involve rolling restarts?

Reply:
Hm, I think we could support that in a simpler way:
- To migrate credentials, first update all of the config files to use the new credentials
- Add a new API endpoint to the basic auth authenticator/escalator/authorizer etc. that triggers a re-read of properties from the config files, user can call this to update the active credentials after the config file update

What do you think about the above? @gianm @nishantmonu51

Reply:
@mohammadjkhan 

I spoke briefly with @gianm about this, he suggested having any usernames and passwords that would need to be potentially updated use a  `PasswordProvider`, where you could create a custom extension that handles any update logic needed. That sounds good to me, would that work for you?

Reply:
+1 on having a {{PasswordProvider}} that can handle password updates without druid restart. It would be useful for non-ldap users too and could potentially improve credentials management.

Reply:
also, IMO, the work on {{PasswordProvider}} can be done in a separate PR and need not block this one.

Reply:
+1 on @nishantmonu51 @jon-wei @gianm feedback on introducing the PasswordProvider approach for rolling credential update that wouldn't require restarting the cluster. I'll go ahead and remove this rolling credential update feature from this PR, and will create a separate ticket/PR that addresses this issue using PasswordProvider approach. Thanks

Reply:
Fixed, removed rolling credential update feature